src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,252 @@ func (b baseBootstrapper) String() string {
+		return bootstrap.NamespaceResults{}, err
 	}
 
-	fulfilledRanges := result.ShardTimeRanges{}
-	fulfilledRanges.AddRanges(currStatus.fulfilled)
-	fulfilledRanges.AddRanges(nextStatus.fulfilled)
-	unfulfilled := totalRanges.Copy()
-	unfulfilled.Subtract(fulfilledRanges)
+	// Unless next bootstrapper is required, this is the final results.
+	finalResults := currResults
 
-	step.mergeResults(unfulfilled)
+	// If there are some time ranges the current bootstrapper could not fulfill,
+	// that we can attempt then pass it along to the next bootstrapper.
+	if next.Namespaces.Len() > 0 {
+		nextResults, err := b.next.Bootstrap(next)
+		if err != nil {
+			return bootstrap.NamespaceResults{}, err
+		}
 
-	unattemptedNextRanges := currRanges.Copy()
-	unattemptedNextRanges.Subtract(currStatus.fulfilled)
-	if !nextAttempted {
-		// If we have never attempted the next time range then we want to also
-		// attempt the ranges we didn't even try to attempt
-		unattemptedNextRanges.AddRanges(nextRanges)
+		// Now merge the final results.
+		for _, elem := range nextResults.Results.Iter() {
+			id := elem.Key()
+			currNamespace := elem.Value()
+
+			finalResult, ok := finalResults.Results.Get(id)
+			if !ok {
+				return bootstrap.NamespaceResults{},
+					fmt.Errorf("expected result for namespace: %s", id.String())
+			}
+
+			// NB(r): Since we originally passed all unfulfilled ranges to the
+			// next bootstrapper, the final unfulfilled is simply what it could
+			// not fulfill.
+			finalResult.DataResult.SetUnfulfilled(currNamespace.DataResult.Unfulfilled().Copy())
+			if currNamespace.Metadata.Options().IndexOptions().Enabled() {
+				finalResult.IndexResult.SetUnfulfilled(currNamespace.IndexResult.Unfulfilled().Copy())
+			}
+
+			// Map is by value, set the result altered struct.
+			finalResults.Results.Set(id, finalResult)
+		}
 	}
 
-	// If there are some time ranges the current bootstrapper could not fulfill,
-	// that we can attempt then pass it along to the next bootstrapper.
-	// NB(r): We explicitly do not ask the next bootstrapper to retry ranges
-	// it could not fulfill as it's unlikely to be able to now.
-	if !unattemptedNextRanges.IsEmpty() {
-		nextStatus, nextErr = step.runNextStep(unattemptedNextRanges)
-		if nextErr != nil {
-			return nextErr
+	return finalResults, nil
+}
+
+func (b baseBootstrapper) logSuccessAndDetermineCurrResultsUnfulfilledAndNextBootstrapRanges(
+	requested bootstrap.Namespaces,
+	curr bootstrap.Namespaces,
+	currResults bootstrap.NamespaceResults,
+	baseLogFields []zapcore.Field,
+) (bootstrap.Namespaces, error) {
+	next := bootstrap.Namespaces{
+		Namespaces: bootstrap.NewNamespacesMap(bootstrap.NamespacesMapOptions{}),
+	}
+	for _, elem := range requested.Namespaces.Iter() {
+		id := elem.Key()
+		requestedNamespace := elem.Value()
+
+		currResult, ok := currResults.Results.Get(id)
+		if !ok {
+			return bootstrap.Namespaces{},
+				fmt.Errorf("namespace result not returned by bootstrapper: %v", id.String())
```

Comment:
Is there any chance that this bootstrapper may just not have the given namespace?